app/gui/src/model/suggestion_database.rs

```diff
@@ -27,6 +28,165 @@ pub use example::Example;
+    }
+}
+
+impl From<String> for FreeformPath {
+    fn from(path: String) -> FreeformPath {
+        path.as_str().into()
+    }
+}
+
+impl From<&Entry> for FreeformPath {
+    fn from(entry: &Entry) -> FreeformPath {
+        match entry.kind {
+            Kind::Method => {
+                let mut path = match &entry.self_type {
+                    Some(name) => FreeformPath::from_segments(name.segments()),
+                    None => default(),
+                };
+                path.segments.push(entry.name.clone().into());
+                path
+            }
+            Kind::Module => FreeformPath::from_segments(entry.module.segments()),
+            _ => {
+                let mut path = FreeformPath::from_segments(entry.module.segments());
+                path.segments.push(entry.name.clone().into());
+                path
+            }
+        }
+    }
+}
+
+
+
+// ===========================
+// === FreeformPathToIdMap ===
+// ===========================
+
+#[derive(Clone, Debug, Default)]
+struct FreeformPathToIdMap {
+    tree: RefCell<ensogl::data::HashMapTree<PathSegment, Option<entry::Id>>>,
+}
+
+impl FreeformPathToIdMap {
+    fn warn_if_exists_and_set(
+        &self,
+        path: impl Into<FreeformPath>,
+        id: entry::Id,
+        logger: &Logger,
+    ) {
+        let path = path.into();
+        let existed = self.check_if_exists_and_set(&path, id);
+        if existed {
+            let path = path.segments;
+            warning!(logger, "An existing id at {path:?} was overwritten with {id}.");
+        }
+    }
+
+    fn warn_if_absent_and_remove(&self, path: impl Into<FreeformPath>, logger: &Logger) {
+        let path = path.into();
+        let existed = self.check_if_exists_and_remove(&path);
+        if !existed {
+            let path = path.segments;
+            let msg =
+                format!("When removing an id at {path:?} some id was expected but none was found.");
+            warning!(logger, "{msg}");
+        }
+    }
+
+    fn check_if_exists_and_set(&self, path: &FreeformPath, id: entry::Id) -> bool {
+        let mut tree = self.tree.borrow_mut();
+        let node = &mut tree.get_or_create_node(&path.segments);
+        let old_value = std::mem::replace(&mut node.value, Some(id));
+        old_value.is_some()
+    }
+
+    fn check_if_exists_and_remove(&self, path: &FreeformPath) -> bool {
+        /// Clears the value at `path` in `node` and returns `true` as the first boolean if the
+        /// value was `Some(_)` before the clearing. Returns `true` as the second boolean if the
+        /// subtree at `node` only contains `None` values after the clearing, or returns `false`
+        /// and removes the first node on `path` that only contains `None` values in its subtree.
+        ///
+        /// The word "empty" is used in the function as a shorthand for "contains only `None`
+        /// values".
+        fn check_if_exists_and_clear_value_and_prune_empty_subtree(
```

Comment:
Such method of "remove leaf and all to-be-empty parents" could be implemented on HasMapTree directly.

Reply:
I moved the function to `HashMapTree` and refactored it significantly. As I mentioned in another comment, I don't currently see a way to further simplify it without reducing its efficiency.